### PR TITLE
Add support for Arm64EC to the Standard Library

### PR DESCRIPTION
**Pull request by dpaoliello**
Adds the final pieces so that the standard library can be built for arm64ec-pc-windows-msvc (initially added in #119199)

* Bumps `windows-sys` to 0.56.0, which adds support for Arm64EC.
* Correctly set the `isEC` parameter for LLVM's `writeArchive` function.
* Add `#![feature(asm_experimental_arch)]` to library crates where Arm64EC inline assembly is used, as it is currently unstable.

### CODE REVIEW

**Comment by workingjubilee** [comment]
...I'm confused why such a pervasive library change assigned someone from docs?

**Comment by dpaoliello** [comment]
> ...I'm confused why such a pervasive library change assigned someone from docs? Rustbot? You okay over there?

I did change some docs, so 🤷

File: src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md [hunk 1]
```diff
@@ -11,48 +11,66 @@ applications on AArch64 Windows 11. See <https://learn.microsoft.com/en-us/windo
...
+  - For example, <https://github.com/rust-lang/portable-simd/commit/ca4033f49b1f6019561b8b161b4097b4a07f2e1b>
+    and <https://github.com/rust-lang/stdarch/commit/166ef7ba22b6a1d908d4b29a36e68ceca324808a>.
+* Assembly for AArch64 might be reusable for Arm64EC, but there are many
+  caveats. For full details see [Microsoft's documentation on the Arm64EC ABI](https://learn.microsoft.com/en-us/windows/arm/arm64ec-abi)
+  but in brief:
+  - Arm64EC uses a restricted set of registers versus AArch64.
+  - Arm64EC uses a different name mangling scheme than AArch64.
+  - Arm64EC requires entry and exit thunks be generated for some functions.
+  - Indirect calls must be done via a call checker.
+  - Contol Flow Guard and stack checks use different functions than AArch64.
```
Comment (GuillaumeGomez) [inline src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md:48]:
```suggestion
  - Control Flow Guard and stack checks use different functions than AArch64.
```

File: src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md [hunk 2]
```diff
@@ -11,48 +11,66 @@ applications on AArch64 Windows 11. See <https://learn.microsoft.com/en-us/windo
...
+    addons into your application), you should use the same path as the x86_64
+    version of your application, not the AArch64 path (since Arm64EC (i.e.,
+    x86_64) processes cannot load native AArch64 DLLs).
+* Arm64EC uses AArch64 intrinsics.
+  - For example, <https://github.com/rust-lang/portable-simd/commit/ca4033f49b1f6019561b8b161b4097b4a07f2e1b>
+    and <https://github.com/rust-lang/stdarch/commit/166ef7ba22b6a1d908d4b29a36e68ceca324808a>.
+* Assembly for AArch64 might be reusable for Arm64EC, but there are many
+  caveats. For full details see [Microsoft's documentation on the Arm64EC ABI](https://learn.microsoft.com/en-us/windows/arm/arm64ec-abi)
+  but in brief:
+  - Arm64EC uses a restricted set of registers versus AArch64.
```
Comment (GuillaumeGomez) [inline src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md:44]: "versus" sounds a bit weird in this context, no?

**Comment by GuillaumeGomez** [comment]
Apart from my nits, docs changes look good. :wink: 

However, please assign someone else for the final approval.

File: src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md (hunk 2 above)
Comment (workingjubilee) [reply src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md:44]: subset, maybe?

**Comment by dpaoliello** [comment]
r? libs

File: src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md (hunk 2 above)
Comment (dpaoliello) [reply src/doc/rustc/src/platform-support/arm64ec-pc-windows-msvc.md]: Fair enough

**Comment by ChrisDenton** [comment]
`windows_sys.rs` is a generated file so any manual changes will be overwritten. See https://github.com/rust-lang/rust/blob/master/library/std/src/sys/pal/windows/c/README.md

You can override definitions by putting them in `library/std/src/sys/pal/windows/c.rs` but ideally fixes should be done in [windows-bindgen](https://github.com/microsoft/windows-rs) (well, more likely the [win32metadata project](https://github.com/microsoft/win32metadata) from which bindings are derived).

**Comment by dpaoliello** [comment]
> `windows_sys.rs` is a generated file so any manual changes will be overwritten. See https://github.com/rust-lang/rust/blob/master/library/std/src/sys/pal/windows/c/README.md
> 
> You can override definitions by putting them in `library/std/src/sys/pal/windows/c.rs` but ideally fixes should be done in [windows-bindgen](https://github.com/microsoft/windows-rs) (well, more likely the [win32metadata project](https://github.com/microsoft/win32metadata) from which bindings are derived).

Submitted a PR to add Arm64EC support in windows-rs: <https://github.com/microsoft/windows-rs/pull/2957>

**Comment by dpaoliello** [comment]
@rustbot blocked

Need to update windows-rs, and I'm still seeing failures with raw-dylib that I thought I had fixed in LLVM 18.1.2

**Comment by dpaoliello** [comment]
`windows-sys` 0.56 has support for Arm64EC, so this PR is ready to go
@rustbot ready

**Review by ChrisDenton — APPROVED** [review]
The libs changes look good to me, though I've not tested it.

**Comment by dpaoliello** [comment]
> The libs changes look good to me, though I've not tested it.

Thanks!

Sounds like the only piece left to review is the compiler changes.
r? compiler

**Review by wesleywiser — APPROVED** [review]
Compiler changes look good to me

**Pull request merged**